Copy: Fix conj bit being ignored on type mismatch

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69026
* __->__ #68963
* #68950
* #68949

Differential Revision: [D33064492](https://our.internmc.facebook.com/intern/diff/D33064492)

peterbell10 committed:
Copy: Fix conj bit being ignored on type mismatch

[ghstack-poisoned]

peterbell10 committed:
Update on "Copy: Fix conj bit being ignored on type mismatch"




[ghstack-poisoned]

peterbell10 committed:
Update on "Copy: Fix conj bit being ignored on type mismatch"




[ghstack-poisoned]

anjali411 commented on test/test_torch.py:
```diff
@@ -5609,6 +5609,42 @@ def test_copy_all_dtypes_and_devices(self, device):
+        for dst_dtype, src_dtype in [
+                (torch.float32, torch.float32),
+                (torch.float64, torch.float32),
+                (torch.int64, torch.int32),
```
there's no way to have int tensors with neg bit no?

peterbell10 replied:
Yes this isn't in the public API but since it is possible with the private API I don't think there's any harm in testing it.

anjali411 commented on test/test_torch.py:
```diff
@@ -5609,6 +5609,42 @@ def test_copy_all_dtypes_and_devices(self, device):
+                (torch.float32, torch.float32),
+                (torch.float64, torch.float32),
+                (torch.int64, torch.int32),
+                (torch.complex128, torch.complex64),
```
similarly I don't there's a way to have complex tensors with neg bit right now

anjali411 reviewed (approved):
Thanks for the fix and the elaborate test!

Separately, it could be cool to add these inter-dtype tests to `OpInfos` because then we can test all the other features on these inputs as well. I believe you can already do that by updating the sample inputs method to manually generate these inputs. Maybe adding a kwarg (`cross_dtype_testing`) and automatically generating these inputs could be a cool addition to OpInfos. cc. @mruberry

mruberry commented:
> Thanks for the fix and the elaborate test!
> 
> Separately, it could be cool to add these inter-dtype tests to `OpInfos` because then we can test all the other features on these inputs as well. I believe you can already do that by updating the sample inputs method to manually generate these inputs. Maybe adding a kwarg (`cross_dtype_testing`) and automatically generating these inputs could be a cool addition to OpInfos. cc. @mruberry

Cross-dtype testing is coming soon but cross-dtype AND conj bit testing is not

ngimel commented:
Please no, it will blow up testing time significantly. Few ops that need cross-dtype testing better do it manually.

anjali411 commented:
> Please no, it will blow up testing time significantly. Few ops that need cross-dtype testing better do it manually.

@ngimel Hmm but it would be an opt-in mechanism!

> Cross-dtype testing is coming soon but cross-dtype AND conj bit testing is not

@ngimel wouldn't we run all the tests with cross-dtype inputs as a part of what's coming up? if not, what's the plan?

mruberry commented:
> > Please no, it will blow up testing time significantly. Few ops that need cross-dtype testing better do it manually.
> 
> @ngimel Hmm but it would be an opt-in mechanism!
> 
> > Cross-dtype testing is coming soon but cross-dtype AND conj bit testing is not
> 
> @mruberry wouldn't we run all the tests with cross-dtype inputs as a part of what's coming up? if not, what's the plan?

In https://github.com/pytorch/pytorch/pull/69439 the elementwise binary test suite was extended with a function that spot-checks type promotion, and there are existing type promotion test for elementwise unary and reduction operators. We don't support type promotion in general, however, so there's not a great universal type promotion or multi-dtype test we can run.

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33064492).

peterbell10 committed:
Update on "Copy: Fix conj bit being ignored on type mismatch"


Differential Revision: [D33064492](https://our.internmc.facebook.com/intern/diff/D33064492)

[ghstack-poisoned]

peterbell10 committed:
Update on "Copy: Fix conj bit being ignored on type mismatch"


Differential Revision: [D33064492](https://our.internmc.facebook.com/intern/diff/D33064492)

[ghstack-poisoned]

peterbell10 committed:
Update on "Copy: Fix conj bit being ignored on type mismatch"


Differential Revision: [D33064492](https://our.internmc.facebook.com/intern/diff/D33064492)

[ghstack-poisoned]

peterbell10 committed:
Rebase and remove as_strided x-fail on "Copy: Fix conj bit being ignored on type mismatch"


Differential Revision: [D33064492](https://our.internmc.facebook.com/intern/diff/D33064492)

[ghstack-poisoned]

peterbell10 committed:
Update on "Copy: Fix conj bit being ignored on type mismatch"


Differential Revision: [D33064492](https://our.internmc.facebook.com/intern/diff/D33064492)

[ghstack-poisoned]

peterbell10 committed:
Update on "Copy: Fix conj bit being ignored on type mismatch"


Differential Revision: [D33064492](https://our.internmc.facebook.com/intern/diff/D33064492)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33064492).
